File: server/src/services/dashboard.service.ts
import { Employee } from '../models/employee.model.js';
import { Department } from '../models/department.model.js';
import { Leave } from '../models/leave.model.js';
import { Attendance } from '../models/attendance.model.js';
import { Payroll } from '../models/payroll.model.js';

export class DashboardService {
  async getDashboard(user: any) {
    if (user.role === 'admin') return this.adminDashboard();
    if (user.role === 'manager') return this.managerDashboard(user.id);
    return this.employeeDashboard(user.id);
  }

  private async adminDashboard() {
    const today = new Date(); today.setHours(0, 0, 0, 0);
    const now = new Date();

    const [totalEmployees, totalDepartments, pendingLeaves, presentToday, payrollResult] = await Promise.all([
      Employee.countDocuments(),
      Department.countDocuments(),
      Leave.countDocuments({ status: 'pending' }),
      Attendance.countDocuments({ date: today, status: { $in: ['present', 'late'] } }),
      Payroll.aggregate([
        { $match: { month: now.getMonth() + 1, year: now.getFullYear() } },
        { $group: { _id: null, total: { $sum: '$netPay' } } },
      ]),
    ]);

    const departmentStats = await Employee.aggregate([
      { $group: { _id: '$departmentId', count: { $sum: 1 } } },
      { $match: { _id: { $ne: null } } },
      { $lookup: {
          from: 'departments',
          let: { deptId: '$_id' },
          pipeline: [
            { $addFields: { strId: { $toString: '$_id' } } },
            { $match: { $expr: { $eq: ['$strId', '$$deptId'] } } },
          ],
          as: 'dept',
        },
      },
      { $unwind: { path: '$dept', preserveNullAndEmptyArrays: true } },
      { $project: { name: '$dept.name', count: 1, _id: 0 } },
    ]);

    const recentLeaves = await Leave.find()
      .populate('employeeId').sort({ createdAt: -1 }).limit(5);

    return {
      totalEmployees, totalDepartments, pendingLeaves,
      presentToday, monthlyPayroll: payrollResult[0]?.total || 0,
      departmentStats, recentLeaves,
    };
  }

  private async managerDashboard(userId: string) {
    const emp = await Employee.findOne({ userId });
    if (!emp) return {};
    const dept = await Department.findById(emp.departmentId);
    const today = new Date(); today.setHours(0, 0, 0, 0);
    const now = new Date();

    const deptEmployeeIds = (await Employee.find({ departmentId: emp.departmentId })
      .select('_id').lean()).map(e => e._id);

    const [totalEmployees, pendingLeaves, presentToday, payrollResult] = await Promise.all([
      Employee.countDocuments({ departmentId: emp.departmentId }),
      Leave.countDocuments({ status: 'pending', employeeId: { $in: deptEmployeeIds } }),
      Attendance.countDocuments({ date: today, status: { $in: ['present', 'late'] }, employeeId: { $in: deptEmployeeIds } }),
      Payroll.aggregate([
        { $match: { month: now.getMonth() + 1, year: now.getFullYear(), employeeId: { $in: deptEmployeeIds } } },
        { $group: { _id: null, total: { $sum: '$netPay' } } },
      ]),
    ]);

    return {
      departmentName: (dept as any)?.name || 'N/A',
      totalEmployees, pendingLeaves, presentToday,
      departmentPayroll: payrollResult[0]?.total || 0,
    };
  }

  private async employeeDashboard(userId: string) {
    const emp = await Employee.findOne({ userId });
    if (!emp) return {};

    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const endOfMonth = new Date(now.getFullYear(), now.getMonth() + 1, 0);

    const [leaves, attendance, lastPayroll, upcomingLeaves] = await Promise.all([
      Leave.find({ employeeId: emp._id }),
      Attendance.find({ employeeId: emp._id, date: { $gte: startOfMonth, $lte: endOfMonth } }),
      Payroll.findOne({ employeeId: emp._id }).sort({ year: -1, month: -1 }),
      Leave.find({ employeeId: emp._id, startDate: { $gte: now }, status: 'approved' }).sort({ startDate: 1 }).limit(3),
    ]);

    const myLeaves: any = { pending: 0, approved: 0, rejected: 0 };
    leaves.forEach(l => { myLeaves[l.status]++; });

    const myAttendance: any = { present: 0, late: 0, absent: 0, halfDay: 0, totalDays: attendance.length };
    attendance.forEach(a => {
      if (a.status === 'half-day') myAttendance.halfDay++;
      else myAttendance[a.status]++;
    });

    return { myLeaves, myAttendance, lastPayroll, upcomingLeaves };
  }
}

File: server/src/main/java/com/hrmanagement/dashboard/service/DashboardService.java
package com.hrmanagement.dashboard.service;

import com.hrmanagement.attendance.repository.AttendanceRepository;
import com.hrmanagement.department.entity.Department;
import com.hrmanagement.department.repository.DepartmentRepository;
import com.hrmanagement.dashboard.dto.AdminDashboardResponse;
import com.hrmanagement.dashboard.dto.EmployeeDashboardResponse;
import com.hrmanagement.dashboard.dto.ManagerDashboardResponse;
import com.hrmanagement.employee.entity.Employee;
import com.hrmanagement.employee.repository.EmployeeRepository;
import com.hrmanagement.leave.entity.Leave;
import com.hrmanagement.leave.repository.LeaveRepository;
import com.hrmanagement.payroll.entity.Payroll;
import com.hrmanagement.payroll.repository.PayrollRepository;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.*;

@Service
public class DashboardService {

    private final EmployeeRepository employeeRepository;
    private final DepartmentRepository departmentRepository;
    private final LeaveRepository leaveRepository;
    private final AttendanceRepository attendanceRepository;
    private final PayrollRepository payrollRepository;

    public DashboardService(EmployeeRepository employeeRepository,
                            DepartmentRepository departmentRepository,
                            LeaveRepository leaveRepository,
                            AttendanceRepository attendanceRepository,
                            PayrollRepository payrollRepository) {
        this.employeeRepository = employeeRepository;
        this.departmentRepository = departmentRepository;
        this.leaveRepository = leaveRepository;
        this.attendanceRepository = attendanceRepository;
        this.payrollRepository = payrollRepository;
    }

    public Object getDashboard(String userRole, String userId) {
        return switch (userRole) {
            case "admin" -> adminDashboard();
            case "manager" -> managerDashboard(userId);
            default -> employeeDashboard(userId);
        };
    }

    private AdminDashboardResponse adminDashboard() {
        LocalDate today = LocalDate.now();
        LocalDate now = today;
        int currentMonth = now.getMonthValue();
        int currentYear = now.getYear();

        long totalEmployees = employeeRepository.count();
        long totalDepartments = departmentRepository.count();
        long pendingLeaves = leaveRepository.countByStatus("pending");
        long presentToday = attendanceRepository.countByDateAndStatusIn(today, List.of("present", "late"));
        BigDecimal monthlyPayroll = payrollRepository.sumNetPayByMonthAndYear(currentMonth, currentYear);

        List<Map<String, Object>> departmentStats = new ArrayList<>();
        List<Department> departments = departmentRepository.findAll();
        for (Department dept : departments) {
            long count = employeeRepository.countByDepartmentId(dept.getId());
            departmentStats.add(Map.of("name", dept.getName(), "count", count));
        }

        List<Leave> recentLeavesList = leaveRepository.findAll(
                org.springframework.data.domain.PageRequest.of(0, 5, org.springframework.data.domain.Sort.by(org.springframework.data.domain.Sort.Direction.DESC, "createdAt"))
        ).getContent();
        List<Map<String, Object>> recentLeaves = recentLeavesList.stream().map(l -> {
            Map<String, Object> m = new HashMap<>();
            m.put("_id", l.getId());
            m.put("type", l.getType());
            m.put("status", l.getStatus());
            m.put("startDate", l.getStartDate());
            m.put("endDate", l.getEndDate());
            if (l.getEmployeeId() != null) {
                m.put("employeeId", Map.of(
                        "_id", l.getEmployeeId().getId(),
                        "firstName", l.getEmployeeId().getFirstName(),
                        "lastName", l.getEmployeeId().getLastName()
                ));
            }
            return m;
        }).toList();

        return new AdminDashboardResponse(totalEmployees, totalDepartments, pendingLeaves,
                presentToday, monthlyPayroll, departmentStats, recentLeaves);
    }

    private ManagerDashboardResponse managerDashboard(String userId) {
        Optional<Employee> mgrEmp = employeeRepository.findByUserId(userId);
        if (mgrEmp.isEmpty() || mgrEmp.get().getDepartmentId() == null) {
            return new ManagerDashboardResponse("N/A", 0, 0, 0, BigDecimal.ZERO);
        }

        Employee emp = mgrEmp.get();
        Department dept = emp.getDepartmentId();
        LocalDate today = LocalDate.now();
        int currentMonth = today.getMonthValue();
        int currentYear = today.getYear();

        List<String> deptEmpIds = employeeRepository.findByDepartmentId(dept.getId())
                .stream().map(Employee::getId).toList();

        long totalEmployees = deptEmpIds.size();
        long pendingLeaves = leaveRepository.countByStatus("pending");
        long presentToday = attendanceRepository.countByEmployeeIdInAndDateAndStatusIn(deptEmpIds, today, List.of("present", "late"));
        BigDecimal departmentPayroll = payrollRepository.sumNetPayByEmployeeIdsAndMonthAndYear(deptEmpIds, currentMonth, currentYear);

        return new ManagerDashboardResponse(dept.getName(), totalEmployees, pendingLeaves, presentToday, departmentPayroll);
    }

    private EmployeeDashboardResponse employeeDashboard(String userId) {
        Optional<Employee> empOpt = employeeRepository.findByUserId(userId);
        if (empOpt.isEmpty()) {
            return new EmployeeDashboardResponse(Map.of(), Map.of(), null, List.of());
        }

        Employee emp = empOpt.get();
        LocalDate now = LocalDate.now();
        LocalDate startOfMonth = now.withDayOfMonth(1);
        LocalDate endOfMonth = now.withDayOfMonth(now.lengthOfMonth());

        List<Leave> allLeaves = leaveRepository.findByEmployeeId(emp.getId(),
                org.springframework.data.domain.PageRequest.of(0, 1000)).getContent();

        Map<String, Long> myLeaves = new HashMap<>();
        myLeaves.put("pending", 0L);
        myLeaves.put("approved", 0L);
        myLeaves.put("rejected", 0L);
        for (Leave l : allLeaves) {
            myLeaves.merge(l.getStatus(), 1L, Long::sum);
        }

        List<com.hrmanagement.attendance.entity.Attendance> monthAttendance =
                attendanceRepository.findByEmployeeIdAndDateBetween(emp.getId(), startOfMonth, endOfMonth);

        Map<String, Long> myAttendance = new HashMap<>();
        myAttendance.put("present", 0L);
        myAttendance.put("late", 0L);
        myAttendance.put("absent", 0L);
        myAttendance.put("halfDay", 0L);
        myAttendance.put("totalDays", (long) monthAttendance.size());
        for (com.hrmanagement.attendance.entity.Attendance a : monthAttendance) {
            switch (a.getStatus()) {
                case "present" -> myAttendance.merge("present", 1L, Long::sum);
                case "late" -> myAttendance.merge("late", 1L, Long::sum);
                case "absent" -> myAttendance.merge("absent", 1L, Long::sum);
                case "half-day" -> myAttendance.merge("halfDay", 1L, Long::sum);
            }
        }

        Payroll lastPayroll = payrollRepository.findByEmployeeId(emp.getId(),
                org.springframework.data.domain.PageRequest.of(0, 1, org.springframework.data.domain.Sort.by(org.springframework.data.domain.Sort.Direction.DESC, "year", "month")))
                .stream().findFirst().orElse(null);

        List<Leave> upcomingLeaves = leaveRepository.findByEmployeeId(emp.getId(),
                org.springframework.data.domain.PageRequest.of(0, 100)).getContent()
                .stream()
                .filter(l -> "approved".equals(l.getStatus()) && !l.getStartDate().isBefore(now))
                .sorted(Comparator.comparing(Leave::getStartDate))
                .limit(3)
                .toList();

        List<Map<String, Object>> upcomingLeavesMapped = upcomingLeaves.stream().map(l -> {
            Map<String, Object> m = new HashMap<>();
            m.put("_id", l.getId());
            m.put("type", l.getType());
            m.put("startDate", l.getStartDate());
            m.put("endDate", l.getEndDate());
            m.put("status", l.getStatus());
            return m;
        }).toList();

        return new EmployeeDashboardResponse(myLeaves, myAttendance, lastPayroll, upcomingLeavesMapped);
    }
}
